feat: validate Edge Functions manifest

jackiewmacharia opened the pull request:
🎉 Thanks for submitting a pull request! 🎉

#### Summary

Fixes https://github.com/netlify/pillar-runtime/issues/289

<!--
Explain the **motivation** for making this change. What existing problem does the pull request solve and how?
-->
To test this:
- Add a sleep function [at the beginning](https://github.com/netlify/build/blob/feat/validate-edge-manifest/packages/build/src/plugins_core/edge_functions/validate_manifest/validate_edge_functions_manifest.js#L72) of the manifest validation core step to allow a few seconds to edit and break the generated manifest.json file in a project `.netlify/edge-functions-dist/manifest.json`:
```
const sleep = (ms) => new Promise((res) => {
  setTimeout(res, ms);
})
await sleep(15000)

```

Screenshots:
- Valid manifest:
<img width="806" alt="valid-json" src="https://user-images.githubusercontent.com/16131515/173890906-51448b96-9709-4b6e-a491-39da461b5bae.png">

- missing-property:
<img width="816" alt="invalid-prop" src="https://user-images.githubusercontent.com/16131515/173892956-f5e411e5-a951-4a6c-ae1c-d934d7b81143.png">

- invalid json syntax:
<img width="1113" alt="malformed-manifest" src="https://user-images.githubusercontent.com/16131515/173887796-0e891683-9c1b-459b-aa2d-cfb1acf42443.png">

---

For us to review and ship your PR efficiently, please perform the following steps:

- [ ] Open a [bug/issue](https://github.com/netlify/build/issues/new/choose) before writing your code 🧑‍💻. This ensures we can discuss the changes and get feedback from everyone that should be involved. If you\`re fixing a typo or something that\`s on fire 🔥 (e.g. incident related), you can skip this step.
- [ ] Read the [contribution guidelines](../CONTRIBUTING.md) 📖. This ensures your code follows our style guide and
      passes our tests.
- [ ] Update or add tests (if any source code was changed or added) 🧪
- [ ] Update or add documentation (if features were changed or added) 📝
- [ ] Make sure the status checks below are successful ✅

**A picture of a cute animal (not mandatory, but encouraged)**
🐙

danez commented:
Maybe it would be easier to maintain this feature by using a json schema and [ajv](https://ajv.js.org/). So we would not need to have our own code for the validation, just the json schema. I guess it would be slightly slower though.

eduardoboucas commented:
> Maybe it would be easier to maintain this feature by using a json schema and [ajv](https://ajv.js.org/). So we would not need to have our own code for the validation, just the json schema. I guess it would be slightly slower though.

I think that's a good direction. I don't think the manifest files will get to a size where performance becomes a problem.

jackiewmacharia commented:
@danez @eduardoboucas I've switched it to use ajv and added test steps + screenshots to the PR description. I've done the validation in a separate step because the manifest is created during the bundling step and wanted to make sure it's ready by validation time.

Also, is there a way to intercept the manifest.json file during snapshot tests in [`runFixture`](https://github.com/netlify/build/blob/feat/validate-edge-manifest/packages/build/tests/edge_functions/tests.js#L100) to break the generated manifest for unhappy path tests? 🤔

eduardoboucas commented on packages/build/src/plugins_core/edge_functions/validate_manifest/validate_edge_functions_manifest.js:
```diff
@@ -0,0 +1,100 @@
+// checks if the pattern string starts with ^ and ends with $
+// we define this format in edge-bundler:
+// https://github.com/netlify/edge-bundler/blob/main/src/manifest.ts#L66
+const normalizedPatternRegex = /^\^.*\$$/
```
Our regular expressions need to be compatible with Go, which means not supporting certain features that JS regexes support (like negative lookahead). I imagine that it'll be difficult to check for that with our validation rules, but if we can find a way, that would be fantastic!

eduardoboucas reviewed (commented):
I'll come back to this PR to do a proper review, but left a thought that we should consider for a follow up.

danez commented on packages/build/tests/edge_functions/snapshots/tests.js.md:
```diff
@@ -420,6 +488,28 @@ Generated by [AVA](https://avajs.dev).
+        }␊
+      ],␊
+      "bundler_version": "1.0.0"␊
+    }␊
```
Should we always log the manifest? I wonder if this is useful for the user at all if it is valid.

eduardoboucas replied:
I don't think we should.

danez reviewed (commented):
> Also, is there a way to intercept the manifest.json file during snapshot tests in [runFixture](https://github.com/netlify/build/blob/feat/validate-edge-manifest/packages/build/tests/edge_functions/tests.js#L100) to break the generated manifest for unhappy path tests? 🤔

I'm not sure about this and it sounds tricky to do, what I can think of is creating a (unit) test that tests the new step in isolation.

eduardoboucas commented:
> Also for the regex I don't know of anything that validates regexp.

I suppose we could use something like https://www.npmjs.com/package/regexp-tree to parse the regex and look for nodes like `kind: "Lookahead"`, which we know are not supported. That feels like a bigger lift though, so it's fine to ship this as is and capture that work in an future for future consideration.

danez reviewed (commented):
Lets also create a follow-up issue, to look into creating a test for this. This wasn't easily possible because we would need to somehow make the manifest generate invalid, right? But we maybe should at least try to look into it again at some point.

jackiewmacharia commented:
@danez about creating a new issue, I've added some unit tests to this, that should suffice, right? @eduardoboucas this might need your approval as well to reflect requested changes as done.
